### PR TITLE
merge the ResolveJwksURI into refresh() function

### LINKED ISSUE #30261: failed to fetch jwks_uri on first attempt does not recover
**Bug description**
We setup istio with requestauthentication resource to validate jwt tokens. The issuer is a URL which causes istiod to try to the OIDC discovery of the well known endpoint to retrieve the JWKS. If an error occurs in the first attempt during the istiod initialization phase it won't recover ever. This is because the failed attempt is stored also for upcoming istiod attempts to resolve the JWKS.

**Logs**
```
2020-12-31T14:44:52.737506Z	warn	model	Failed to get jwks_uri for issuer "https://$domain/oauth2/default": Get "https://$domain/oauth2/default/.well-known/openid-configuration": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-12-31T14:44:52.737482Z	error	model	Failed to fetch jwks_uri from "https://$domain/oauth2/default/.well-known/openid-configuration": Get "https://$domain/oauth2/default/.well-known/openid-configuration": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

2020-12-31 16:02:32	0.0.0.0_8080: Issuer 'https://$domain/oauth2/default' in jwt_authn config has invalid local jwks: Jwks RSA [n] or [e] field is missing or has a parse error
2020-12-31T15:02:31.992986Z	error	Failed to fetch jwt public key from "": Get "": unsupported protocol scheme ""
2020-12-31T15:02:31.992972Z	error	model	Failed to fetch public key from "": Get "": unsupported protocol scheme ""
2020-12-31T15:02:31.991866Z	error	Failed to fetch jwt public key from "": Get "": unsupported protocol scheme ""
2020-12-31T15:02:31.991851Z	error	model	Failed to fetch public key from "": Get "": unsupported protocol scheme ""
```

In the istiod code here is where the problem starts:
https://github.com/istio/istio/blob/080374a27dd16646be2649ab377af49183667ac1/pilot/pkg/model/jwks_resolver.go#L261

```go	
       body, err := r.getRemoteContentWithRetry(issuer+openIDDiscoveryCfgURLSuffix, networkFetchRetryCountOnMainFlow)
	if err != nil {
		log.Errorf("Failed to fetch jwks_uri from %q: %v", issuer+openIDDiscoveryCfgURLSuffix, err)
		return "", err
	}
```

Here I think it should either retry or bail out (panic) because it will store the empty string on error and will result in the errors you can see above.
https://github.com/istio/istio/blob/080374a27dd16646be2649ab377af49183667ac1/pilot/pkg/model/jwks_resolver.go#L214

[ ] Docs
[x] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Steps to reproduce the bug**
Reproduce it with having a first attempt failure as described above.

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

```
$ kubectl version --short
Client Version: v1.18.6
Server Version: v1.18.9-eks-d1db3c
$ istioctl version
client version: 1.8.1
control plane version: 1.8.1
data plane version: 1.8.1 (6 proxies)
```

**How was Istio installed?**
istioctl

**Environment where the bug was observed (cloud vendor, OS, etc)**
AWS eks but it's not environment specific.

### PR DESCRIPTION
**Pull request by domcyrus**
As suggested by @yangminzhu this PR merges the ResolveJwksURI to the refresh() function. Fixes #30261 

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request.

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: releasenotes/notes/30261.yaml
Comment (yangminzhu) [inline releasenotes/notes/30261.yaml:8]:
```
**Added** `istiod` JWT public key refresh job will now retry the failed fetch of the `jwks_uri` with exponential backoff.
```

File: pilot/pkg/model/jwks_resolver.go
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:346]: Following my comment in #30407, I think we should add some limit to this, maybe exponential backoff.

cc @incfly and @howardjohn if you have other suggestions.

File: pilot/pkg/model/jwks_resolver_test.go
Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver_test.go:71]: This was only an initial test I've started with. It is not used at all. I've removed it.

File: pilot/pkg/model/jwks_resolver_test.go
Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver_test.go:384]: I didn't want to access `r.refreshInterval` directly because it would introduce a data race (refresher in separate goroutine is writing and this test in another goroutine is reading). That said I also think the whole test is a bit tricky as we can't directly observe the refresh interval and also cannot directly control it.

@yangminzhu Do you think we could change `JwksResolver` adding a RW mutex which would make this test much easier. With adding the RW mutex we could then also consider swapping out the `sync.Map` and use a normal map guarded by the RW mutex. I think the runtime cost for the RW mutex in this case should be negligible and the more ergonomic use of a "normal" map should out pay the runtime cost.

```golang
// JwksResolver is resolver for jwksURI and jwt public key.
type JwksResolver struct {
	mu sync.RWMutex

	// Callback function to invoke when detecting jwt public key change.
	PushFunc func()

	// cache for JWT public key.
	keyEntries map[jwtKey]jwtPubKeyEntry
```

File: releasenotes/notes/30261.yaml
Comment (domcyrus) [reply releasenotes/notes/30261.yaml:8]: Thanks 🙇

File: pilot/pkg/model/jwks_resolver_test.go
Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver_test.go]: I've rewritten the unit test for the refresh behavior with using a lock on `r.refreshInterval` and calling refresh directly instead of passively observe it.

File: pilot/pkg/model/jwks_resolver.go
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:245]: I realize this is not made worse by your PR, but I am actually pretty concerned about sending these network calls (with retry) on the XDS generation. Not trying to block this PR, but we may need to investigate this. I have concerns about how it will behave at scale

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:245]: As the `networkFetchRetryCountOnMainFlow` is set to 0 anyway we could also just remove the immediate retries and rely on the refresh mechanism. The refreshs are usually happening every 20 minutes and with the change of this PR on error they are happening 1m, 2m, 4m, 8m, 16m, 20m, 20m...

File: pilot/pkg/model/jwks_resolver.go
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver.go:447]: are we setting more aggressive retry interval for ALL endpoints if a single one fails?

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:447]: Yes you are right... This isn't that great. I can alter the code to only refresh more aggressively the failing endpoints. I was also thinking about if each endpoint should be refreshed with its own interval (spread them with some random time at init) so that they do not synchronously update all at once e.g. if there was some network issue which affected all endpoints. How does that sound? Definitely would need some more work for this but might be better ;)

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:447]: @howardjohn The `PushFunc` is only invoked if there is a change of the key. Even in the error case it will not call the `PushFunc` unless the already cached key is getting evicted (currently 24h) or there is a change of the key. I'm not so familiar with the rotation of jwks_uri but to me it looks like it is not an event which happens often e.g. okta rotates 4 times a year. Therefore I'm curios in what scenario this can be an issue maybe I'm missing something.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:447]: That said shall I add a unittest to show how often the `PushFunc` is getting called so that I can show that it doesn't get called if not needed? @yangminzhu What do you think about this?

File: pilot/pkg/model/jwks_resolver.go
Comment (yangminzhu) [reply pilot/pkg/model/jwks_resolver.go:245]: the current way is to block the xDS a little bit for the network request (the fetched jwks will be cached) and then push the jwks within the same single xDS response, the same result will be cached and reused for all pods, how is this related to the scale of pods?

The other way is to let the separate refresh goroutine to fetch and push asynchronously, but that means an extra xDS push just for the jwks key, and until it completes the request authentication will have no effect.

File: pilot/pkg/model/jwks_resolver.go
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:341]: I really think we should not use the lock if its only purpose is to make the unit test easier, the code gets much more complex and error prone with very small benefits, could you change to something like the below:

```golang
var (
	// JwtPubKeyRefreshIntervalOnFailureResetThreshold is the threshold to reset the refresh interval on failure.
	JwtPubKeyRefreshIntervalOnFailureResetThreshold = 60 * time.Minute
)

func (r *JwksResolver) refresher() {
	// Wake up once in a while and refresh stale items.
	r.refreshTicker = time.NewTicker(r.refreshInterval)
	lastHasError := false
	for {
		select {
		case <-r.refreshTicker.C:
			currentHasError := r.refresh()
			if currentHasError {
				if lastHasError {
					// update to exponential backoff if last time also failed.
					r.refreshInterval *= 2
					if r.refreshInterval > JwtPubKeyRefreshIntervalOnFailureResetThreshold {
						r.refreshInterval = r.refreshIntervalOnFailure
					}
				} else {
					// change to the refreshIntervalOnFailure if failed for the first time.
					r.refreshInterval = r.refreshIntervalOnFailure
				}
			} else {
				// reset the refresh interval if success.
				r.refreshInterval = r.refreshDefaultInterval
			}
			lastHasError = currentHasError
			r.refreshTicker.Stop()
			r.refreshTicker = time.NewTicker(r.refreshInterval)
		case <-closeChan:
			r.refreshTicker.Stop()
			return
		}
	}
}
```

And inside the refresh(), remove the code that updates the `r.refreshInterval`, the refresh() should only care about the single refresh logic, so let's move the backoff code to the outer level.

And in the test, you could do something like the below:

```golang
func TestJwtRefreshIntervalExponentialBackoff(t *testing.T) {
	defaultRefreshInterval := 50*time.Millisecond
	refreshIntervalOnFail := 2*time.Millisecond
	r := newJwksResolver(JwtPubKeyEvictionDuration, defaultRefreshInterval, refreshIntervalOnFail, 1*time.Millisecond)

	ms := startMockServer(t)
	defer ms.Stop()

	// Configures the mock server to return error after the first request.
	ms.ReturnErrorAfterFirstNumHits = 1

	mockCertURL := ms.URL + "/oauth2/v3/certs"
	_, err := r.GetPublicKey("", mockCertURL)
	if err != nil {
		t.Fatalf("GetPublicKey(%q, %+v) fails: expected no error, got (%v)", "", mockCertURL, err)
	}

	time.Sleep(100*time.Millisecond)
	r.Close()

	powerOfTwo := func(n int64) bool {
		return n != 0 && (n & (n - 1) == 0)
	}
	r.refreshInterval.Milliseconds()
	if r.refreshInterval == refreshIntervalOnFail || !powerOfTwo(r.refreshInterval.Milliseconds() / time.Millisecond.Milliseconds()) {
		t.Errorf("refreshInterval not updated with exponential backoff, got %v", r.refreshInterval)
	}
}
```

And you can add another `TestJwtRefreshIntervalRecoverFromFail()`, and set `ReturnErrorForFirstNumHits=3`, sleep a little bit, stop the refresher and then check the `r.refreshInterval == defaultRefreshInterval`.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:341]: Thanks @yangminzhu. This makes totally sense to me.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:341]: @yangminzhu I failed to figure out if `r.GetPublicKey` is getting called periodically. Since this is the only way the cache `r.keyEntries` is getting populated there is currently an issue for which I've added the test `TestJwtRefreshIntervalRecoverFromInitialFailOnFirstHit`. It illustrates the current issue.

I could fix this in that I extend the jwtPubKeyEntry to have a isValid field but I'm curious if that is needed. I think that depends on whether `r.GetPublicKey` is periodically called or not.
```golang
type jwtPubKeyEntry struct {
	pubKey string

	// The last success refreshed time of the pubKey.
	lastRefreshedTime time.Time

	// Cached item's last used time, which is set in GetPublicKey.
	lastUsedTime time.Time

        // isValid indicates whether the public key is valid to use.
        isValid bool
}
```

Using it in `r.GetPublicKey` like this:
```golang
	if jwksURI == "" {
		// Fetch the jwks URI if it is not hardcoded on config.
		var err error
		jwksURI, err = r.resolveJwksURIUsingOpenID(issuer)
		if err != nil {
			log.Errorf("Failed to jwks URI from %q: %v", issuer, err)
	                r.keyEntries.Store(key, jwtPubKeyEntry{
		                lastRefreshedTime: now,
		                lastUsedTime:      now,
                                isValid: false,
	                }
			return "", err
		}
	}
```

This way the key would also get refreshed in `r.refresh` and would trigger it into failure mode to potentially recover more quickly. But as mentioned above I'm not 100% sure if this is needed.

I tried to trace back the pilot xds/lds config generation to see if `r.GetPublicKey` is getting invoked periodically. Since there is key deletion in `r.refresh` like below I guess (hope) that means there must be some sort of period config generation which triggers `r.GetPublicKey` even if there are no external CRD config changes otherwise with the code so far in this PR I fear that we could end up with the exact same issue for which I've opened this PR to fix it ;) Is that the case?
```golang

		// Remove cached item for either of the following 2 situations
		// 1) it hasn't been used for a while
		// 2) it hasn't been refreshed successfully for a while
		// This makes sure 2 things, we don't grow the cache infinitely and also we don't reuse a cached public key
		// with no success refresh for too much time.
		if now.Sub(e.lastUsedTime) >= r.evictionDuration || now.Sub(e.lastRefreshedTime) >= r.evictionDuration {
			log.Infof("Removed cached JWT public key (lastRefreshed: %s, lastUsed: %s) from %q",
				e.lastRefreshedTime, e.lastUsedTime, k.issuer)
			r.keyEntries.Delete(k)
			return true
		}
```

Comment (yangminzhu) [reply pilot/pkg/model/jwks_resolver.go]: `GetPublicKey()` is called when a RequestAuthentication is created or updated, if the jwksUri is fetched successfully without error, the jwksUri will be stored to the `keyEntries` and the refresh job will periodically update the jwksUri in the `keyEntries`.

I think you don't need to worry about that `GetPublicKey()` is not called periodically, but the real problem is when `GetPublickKey()` failed, it won't store the jwksUri to the `keyEntries` which means it won't be refreshed by the periodic job. So I think you could fix this by changing `GetPublicKey()` to always store the jwksUri to the `kenEntries` even if it fails (you can just use an empty pubKey, we called `createFakeJwks()` in policy_applier.go to handle the error case)

That's a very good finding and I think it's not covered in the initial PR, and should be tested with the `TestJwtRefreshIntervalRecoverFromInitialFailOnFirstHit()`, thank you!

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go]: Thanks a lot I've addressed that.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go]: @yangminzhu I've fixed all the outstanding issues and rebased to latest upstream which still looks ok though I don't understand why this PR now shows the commits of pull requests which were approved and are already in master. Anyway I can also open another PR as there were some merge conflicts with #29704 which I did fix.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go]: Just for the record also all the merge checks are now satisfied.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go]: @yangminzhu I've created another branch without the merge conflicts. Please let me know if I should open a PR for this: https://github.com/istio/istio/compare/master...domcyrus:fix-jwks-uri-on-startup3
and close this one. Thanks.

**Comment by yangminzhu** [comment]
@domcyrus Could you fix the PR history to only include your changes? There are 66 commits in the current PR with lots of other people's change. I think you don't need to create a new branch for resolving the conflict, I usually just pull the latest master branch, and rebase my working branch onto it and then push my working branch to github.

**Comment by domcyrus** [comment]
> @domcyrus Could you fix the PR history to only include your changes? There are 66 commits in the current PR with lots of other people's change. I think you don't need to create a new branch for resolving the conflict, I usually just pull the latest master branch, and rebase my working branch onto it and then push my working branch to github.

@yangminzhu ok looks like I just had to `git merge upstream/master` and push. Thanks. I also merged the next conflict: #30014

File: pilot/pkg/model/jwks_resolver_test.go
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver_test.go:393]: remove the first `request`.

**Review by yangminzhu — APPROVED** [review]
This is a really nice improvement, thanks for working on this and addressing all the comments, great job!

**Review by shamsher31 — APPROVED** [review]

**Comment by domcyrus** [comment]
@yangminzhu Thanks a lot for your comments, they were always very helpful!

**Comment by yangminzhu** [comment]
+@howardjohn for review/approval from networking, thanks.

**Review by morvencao — APPROVED** [review]

**Comment by yangminzhu** [comment]
cc @istio/wg-networking-maintainers-pilot for review.

**Comment by ramaraochavali** [comment]
@howardjohn looks like you have reviewed this earlier, can you please approve if this looks good?

File: pilot/pkg/model/jwks_resolver_test.go
Comment (howardjohn) [inline pilot/pkg/model/jwks_resolver_test.go:459]: This test is flaky. example failures:
```
=== RUN   TestJwtPubKeyMetric
2021-02-25T18:06:41.360432Z     info    Successfully serving on http://localhost:37847
2021-02-25T18:06:41.361437Z     error   model   Failed to fetch public key from "http://localhost:37847/oauth2/v3/certs": status 403, Mock server configured to return error until 1 retries
2021-02-25T18:06:41.369742Z     warn    model   Previous JWKs public key JSON is not parseable:
2021-02-25T18:06:41.369797Z     info    model   Updated cached JWT public key from "http://localhost:37847/oauth2/v3/certs"
2021-02-25T18:06:41.421382Z     info    model   Updated cached JWT public key from "http://localhost:37847/oauth2/v3/certs"
    jwks_resolver_test.go:490: GetPublicKey("", [ http://localhost:37847/oauth2/v3/certs]): expected ({ "keys": [ { "kid": "fakeKey1_1", "alg": "RS256", "kty": "RSA", "n": "abc", "e": "def" },
                                { "kid": "fakeKey1_2", "alg": "RS256", "kty": "RSA", "n": "123", "e": "456" } ] }), got ({ "keys": [ { "kid": "fakeKey2_1", "alg": "RS256", "kty": "RSA", "n": "ghi", "e": "lmn" },
                                { "kid": "fakeKey2_2", "alg": "RS256", "kty": "RSA", "n": "789", "e": "1234" } ] })
--- FAIL: TestJwtPubKeyMetric (2.61s)
2021-02-25T18:06:41.482742Z     error   Server failed to serve in "http://localhost:37847": http: Server closed
FAIL

=== RUN   TestJwtPubKeyMetric
2021-02-25T18:06:40.281042Z     info    Successfully serving on http://localhost:37793
2021-02-25T18:06:40.282153Z     error   model   Failed to fetch public key from "http://localhost:37793/oauth2/v3/certs": status 403, Mock server configured to return error until 1 retries
2021-02-25T18:06:40.283383Z     warn    model   Previous JWKs public key JSON is not parseable:
2021-02-25T18:06:40.283429Z     info    model   Updated cached JWT public key from "http://localhost:37793/oauth2/v3/certs"
2021-02-25T18:06:40.334910Z     info    model   Updated cached JWT public key from "http://localhost:37793/oauth2/v3/certs"
    jwks_resolver_test.go:490: GetPublicKey("", [ http://localhost:37793/oauth2/v3/certs]): expected ({ "keys": [ { "kid": "fakeKey1_1", "alg": "RS256", "kty": "RSA", "n": "abc", "e": "def" },
                                { "kid": "fakeKey1_2", "alg": "RS256", "kty": "RSA", "n": "123", "e": "456" } ] }), got ({ "keys": [ { "kid": "fakeKey2_1", "alg": "RS256", "kty": "RSA", "n": "ghi", "e": "lmn" },
                                { "kid": "fakeKey2_2", "alg": "RS256", "kty": "RSA", "n": "789", "e": "1234" } ] })
--- FAIL: TestJwtPubKeyMetric (3.20s)
2021-02-25T18:06:40.403669Z     error   Server failed to serve in "http://localhost:37793": http: Server closed
FAIL
```

See https://github.com/istio/istio/wiki/Test-Flakes for more info

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver_test.go:465]: Thanks for the pointer I've fixed the tests using `retry.UntilOrFail`.

Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver_test.go:465]: @howardjohn I've addressed the flaky tests. Is there anything else needed to get changed?

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
